contracts/PoolManager.sol

```diff
@@ -129,69 +144,54 @@ contract PoolManager is IPoolManager, Owned, NoDelegateCall, ERC1155, IERC1155Re
-    struct LockState {
-        uint256 nonzeroDeltaCount;
-        mapping(Currency currency => int256) currencyDelta;
+    function locksGetter(uint256 i) external view override returns (IPoolManager.Lock memory) {
+        return locks[i];
     }
 
-    /// @dev Represents the state of the locker at the given index. Each locker must have net 0 currencies owed before
-    /// releasing their lock. Note this is private because the nested mappings cannot be exposed as a public variable.
-    mapping(uint256 index => LockState) private lockStates;
-
     /// @inheritdoc IPoolManager
-    function getNonzeroDeltaCount(uint256 id) external view returns (uint256) {
-        return lockStates[id].nonzeroDeltaCount;
+    function locksLength() public view returns (uint256) {
+        return locks.length;
     }
 
     /// @inheritdoc IPoolManager
-    function getCurrencyDelta(uint256 id, Currency currency) external view returns (int256) {
-        return lockStates[id].currencyDelta[currency];
+    function getCurrencyDelta(address locker, Currency currency) external view returns (int256) {
+        return currencyDeltas[locker][currency];
     }
 
     /// @inheritdoc IPoolManager
     function lock(bytes calldata data) external override returns (bytes memory result) {
-        uint256 id = lockedBy.length;
-        lockedBy.push(msg.sender);
+        locks.push(Lock({locker: msg.sender, parentLockIndex: uint96(lockIndex)}));
+        unchecked {
+            lockIndex = uint128(locks.length - 1);
+        }
 
         // the caller does everything in this callback, including paying what they owe via calls to settle
-        result = ILockCallback(msg.sender).lockAcquired(id, data);
+        result = ILockCallback(msg.sender).lockAcquired(data);
 
-        unchecked {
-            LockState storage lockState = lockStates[id];
-            if (lockState.nonzeroDeltaCount != 0) revert CurrencyNotSettled();
+        if (lockIndex == 0) {
+            if (nonzeroDeltaCount != 0) revert CurrencyNotSettled();
+            delete locks;
+        } else {
+            lockIndex = locks[lockIndex].parentLockIndex;
         }
-
-        lockedBy.pop();
     }
 
     function _accountDelta(Currency currency, int128 delta) internal {
         if (delta == 0) return;
 
-        LockState storage lockState = lockStates[lockedBy.length - 1];
-        int256 current = lockState.currencyDelta[currency];
-
+        address locker = locks[lockIndex].locker;
+        int256 current = currencyDeltas[locker][currency];
         int256 next = current + delta;
+
         unchecked {
             if (next == 0) {
-                lockState.nonzeroDeltaCount--;
+                nonzeroDeltaCount--;
             } else if (current == 0) {
-                lockState.nonzeroDeltaCount++;
+                nonzeroDeltaCount++;
             }
         }
 
-        lockState.currencyDelta[currency] = next;
+        currencyDeltas[locker][currency] = next;
     }
 
```

Comment:
`lockIndex == 0` means `locks.length` is 1.
Then what's the purpose of below `for` loop since `length` is 1?